lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,956 @@
+  if (DebugSynthesis)
+    llvm::outs() << "starting synthesis\n";
+  unsigned Rounds = 0;
+  while (true) {
+    Inst *Query = IC.getConst(APInt(1, true));
+    // Each set of concrete inputs is used in a separate copy of the
+    // main WiringQuery. For correct separation of these copies, we must
+    // copy all component input variables that encode data-flow in the
+    // program too (starting with the second concrete input set, see below)
+    for (auto const &InputMap : S) {
+      if (DebugSynthesis) {
+        for (auto const &Input : InputMap) {
+          if (Input.first->Name.find(COMP_INPUT_PREFIX) != std::string::npos)
+            continue;
+          llvm::outs() << "setting input " << Input.first->Name
+                       << " to " << Input.second->Val << "\n";
+        }
+        llvm::outs() << "----\n";
+      }
+      Inst *Copy = getInstCopy(WiringQuery, IC, InputMap);
+      Query = IC.getInst(Inst::And, 1, {Query, Copy});
+    }
+    // Solve the synthesis constraint.
+    // Each solution corresponds to a syntactically distinct and well-formed
+    // straight-line program obtained by composition of given components
+    ModelInsts.clear();
+    ModelVals.clear();
+    InstMapping Mapping(Query, IC.getConst(APInt(1, true)));
+    // Negate the query to get a SAT model.
+    // Don't use original BPCs/PCs, the are useless
+    QueryStr = BuildQuery({}, NewPCs, Mapping, &ModelInsts, /*Negate=*/true);
+    EC = SMTSolver->isSatisfiable(QueryStr, IsSat, ModelInsts.size(),
+                                  &ModelVals, Timeout);
+    // Fail on error or when no valid wiring exists
+    if (EC || !IsSat)
+      return EC;
+
+    if (DebugSynthesis)
+      llvm::outs() << "round: " << Rounds << "\n";
+    Rounds++;
+
+    Inst *Cand = createInstFromModel(ModelInsts, ModelVals, IC);
+    if (!Cand) {
+      llvm::errs() << "synthesis bug: creating inst from a model failed\n";
+      return EC;
+    }
+
+    if (DebugSynthesis) {
+      llvm::outs() << "candidate:\n";
+      ReplacementContext Context;
+      PrintReplacementRHS(llvm::outs(), Cand, Context);
+    }
+
+    // Does the candidate work for all inputs?
+    // Use original BPCs/PCs
+    ModelInsts.clear();
+    ModelVals.clear();
+    InstMapping CandMapping(LHS, Cand);
+    QueryStr = BuildQuery(BPCs, PCs, CandMapping, &ModelInsts, /*Negate=*/false);
+    EC = SMTSolver->isSatisfiable(QueryStr, IsSat, ModelInsts.size(),
+                                  &ModelVals, Timeout);
+    if (EC)
+      return EC;
+
+    // Success
+    if (!IsSat) {
+      if (DebugSynthesis)
+        llvm::outs() << "synthesis succeeded\n";
+      RHS = Cand;
+      return EC;
+    }
+
+    if (DebugSynthesis)
+      llvm::outs() << "didn't work for all inputs, counterexample(s):\n";
+    // Parse input counterexamples from the model
+    std::map<Inst *, Inst *> Replacements;
+    for (unsigned J = 0; J != ModelInsts.size(); ++J) {
```

Comment:
Hmm, it's looks like it's not consistent across the code base (see e.g. lib/Extractor/Solver.cpp, lib/Extractor/KLEEBuilder.cpp).